site/en/blog/full-accessibility-tree/index.md

```diff
@@ -0,0 +1,167 @@
+
+
+### Platforms
+
+Every platform has a different accessibility API and while the shape of the tree is the same across all platforms, the API for interacting with the tree is different, and the names of attributes can differ. DevTools shows Chromium’s internal tree where roles and attributes tend to match those defined in the [ARIA](https://www.w3.org/TR/wai-aria-1.1/) specification.
+
+
+### Multiple frames and site isolation
+
+Since Chromium not only puts the content of every tab in different renderer processes but also [isolates cross-site documents in different renderer processes](https://www.chromium.org/developers/design-documents/site-isolation), we have to connect to each out-of-process child document separately over CDP and fetch its accessibility tree. We then stitch these subtrees together on the frontend to give the illusion of one coherent tree, although they live in different renderer processes in Chromium.
+
+
+### Ignored and uninteresting nodes
+
+We hide some nodes per default: ignored nodes, and nodes with role “generic” with no name. These nodes carry no semantic meaning and, in the case of ignored nodes, are not exposed to assistive technology. We hide these nodes to avoid cluttering the tree view. If we did not, the accessibility tree for most web pages would instead look something like this:
+
+{% Img src="image/gGNVLl64MBQvhaYBXAAuRiOs3r92/K4OmzuAaVcQjFjEZlcZq.png", alt="The new tree view with all nodes shown.", width="800", height="488" %}
+
+The caveat here is that this essentially means that we need to construct yet another tree than what is available on the backend. Say, for example, that we have nodes A, B, C, and X where A has child X and B, and X has child C. If X is an ignored node, we prune X from the tree and instead create a tree where C is a child to A.
+
+{% Img src="image/gGNVLl64MBQvhaYBXAAuRiOs3r92/DqQ4L3DRo6HTg8RIJnwO.png", alt="Diagram showing how we prune the tree.", width="800", height="450" %}
+
+On the frontend, we construct the full tree including ignored nodes and only prune them just before rendering the nodes. We do this for two reasons:
+
+1. **It makes it much simpler to handle node updates from the backend**, since we have the same tree structure on both endpoints. For example, if the node B is removed in the example, we would receive an update for node X (since its children have changed), but if we had pruned that node we would struggle to figure out what to update.
+2. **It ensures that all DOM nodes have a corresponding accessibility node.** When the tree is toggled, we select the node corresponding to the node currently selected in the DOM tree. So for the previous example, if the user toggles the tree while the DOM node corresponding to X is selected, we inject X between the nodes A and B and select X in the tree. This allows the user to inspect the accessibility node for all DOM nodes and help determine why the node is ignored.
```

Comment:
```suggestion
- **It makes it much simpler to handle node updates from the backend**, since we have the same tree structure on both endpoints. For example, if the node B is removed in the example, we would receive an update for node X (since its children have changed), but if we had pruned that node we would struggle to figure out what to update.
- **It ensures that all DOM nodes have a corresponding accessibility node.** When the tree is toggled, we select the node corresponding to the node currently selected in the DOM tree. So for the previous example, if the user toggles the tree while the DOM node corresponding to X is selected, we inject X between the nodes A and B and select X in the tree. This allows the user to inspect the accessibility node for all DOM nodes and help determine why the node is ignored.
```